x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,446 @@
+	if c.VisibilityTimeout < 0 || c.VisibilityTimeout.Hours() > 12 {
+		return errors.New("visibilityTimeout is not defined within the " +
+			"expected bounds")
+	}
+
+	if len(c.QueueURLs) == 0 {
+		return errors.New("no SQS queueURLs are configured")
+	}
+	return nil
+}
+
+func getAWSCredentials(config config) (awssdk.Config, error) {
+	// Check if accessKeyID and secretAccessKey is given from configuration
+	if config.AccessKeyID != "" && config.SecretAccessKey != "" {
+		awsConfig := defaults.Config()
+		awsCredentials := awssdk.Credentials{
+			AccessKeyID:     config.AccessKeyID,
+			SecretAccessKey: config.SecretAccessKey,
+		}
+		if config.SessionToken != "" {
+			awsCredentials.SessionToken = config.SessionToken
+		}
+
+		awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+			Value: awsCredentials,
+		}
+		return awsConfig, nil
+	}
+
+	// If accessKeyID and secretAccessKey is not given, then load from default config
+	// Please see https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html
+	// with more details.
+	return external.LoadDefaultAWSConfig(
+		external.WithSharedConfigProfile(config.ProfileName),
+	)
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.logger.Debugf("Run s3 input with queueURLs: %v", p.config.QueueURLs)
+	visibilityTimeout := int64(p.config.VisibilityTimeout.Seconds())
+
+	for _, queueURL := range p.config.QueueURLs {
+		regionName, err := getRegionFromQueueURL(queueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %v", queueURL)
+			continue
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// receive messages from sqs
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &queueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+		output, err := req.Send(context.Background())
+		if err != nil {
+			p.logger.Error("failed to receive message from SQS:", err)
+			continue
+		}
+
+		if len(output.Messages) == 0 {
+			p.logger.Debug("no message received from SQS:", queueURL)
+			continue
+		}
+
+		// process messages received from sqs, get logs from s3 and create events
+		p.processor(queueURL, output.Messages, visibilityTimeout, svcS3, svcSQS)
+	}
+}
+
+// Stop stops the s3 input
+func (p *Input) Stop() {
+	close(p.close)
+	defer p.outlet.Close()
+	p.logger.Info("Stopping s3 input")
+}
+
+// Wait stops the s3 input.
+func (p *Input) Wait() {
+	p.Stop()
+}
+
+func (p *Input) processor(queueURL string, messages []sqs.Message, visibilityTimeout int64, svcS3 *s3.Client, svcSQS *sqs.Client) {
+	var wg sync.WaitGroup
+	numMessages := len(messages)
+	wg.Add(numMessages)
+
+	// process messages received from sqs
+	for i := range messages {
+		errC := make(chan error)
```

Comment:
This `errC` channel could be used to handle e2e ACK (example here: https://github.com/elastic/beats/pull/12850#issuecomment-513507762). If passed as `Private`, `ACKEvents` would only need to close it to trigger an ACK in SQS. cc @urso